src/twisted/web/test/test_cgi.py

```diff
@@ -363,6 +374,44 @@ class CGIScriptTests(unittest.TestCase):
     """
     Tests for L{twcgi.CGIScript}.
     """
+    def tearDown(self):
+        if getattr(self, 'p', None):
+            return self.p.stopListening()
+
+
+    def startServer(self, cgi):
+        root = resource.Resource()
+        cgipath = util.sibpath(__file__, cgi)
+        root.putChild(b"cgi", PythonScript(cgipath))
+        site = server.Site(root)
+        self.p = reactor.listenTCP(0, site)
+        return self.p.getHost().port
```

Comment:
This style of testing is, unfortunately, prone to spurious failures.  `listenTCP(0, ...)` will fail in some cases.  Also, the platform TCP/IP stack is necessary for testing the application changes here.  Instead, you can exercise all of the logic by moving bytes around in-memory.

I'm not sure what the current best practices in Twisted are for this kind of testing but I know that there is `twisted.test.iosim` and `twisted.internet.testing` with both offer tools that might be of use.

Reply:
This was just copied from CGITests.startServer in the same file.